### PR TITLE
[WIP] Feature/lock screen icon url

### LINKED ISSUE #281: SDL Lock Screen Icon URL Implementation
When a head unit connects it may send an OnSystemRequest notification with RequestType: LOCK_SCREEN_ICON_URL. The URL that is passed should be used by the proxy to retrieve an image. The image will be passed to the developer to be displayed on their lock screen alongside their own logo. The logo will generally represent the make of the vehicle to which the application is connected.

### PR DESCRIPTION
**Pull request by justinjdickow**
[READY FOR REVIEW]
  
  * Handles onSystemRequest notification with RequestType LOCK_SCREEN_ICON_URL
  * Downloads icon at the url in the request
  * Passes icon to developer in onReceivedLockScreenIcon method of SDLProxyListenerProtocol
  * Resolves #281

### CODE REVIEW

File: SmartDeviceLink-iOS/SmartDeviceLink/SDLProxyListener.h [hunk 1]
```diff
@@ -57,6 +57,7 @@
 @class SDLUnregisterAppInterfaceResponse;
 @class SDLUnsubscribeButtonResponse;
 @class SDLUnsubscribeVehicleDataResponse;
+@import UIKit;
```
Comment (joeljfischer) [inline SmartDeviceLink-iOS/SmartDeviceLink/SDLProxyListener.h]: Imports should be above class forwarding

File: SmartDeviceLink-iOS/SmartDeviceLink/SDLProxy.m [hunk 2]
```diff
@@ -592,6 +621,25 @@ - (NSURLSessionDataTask *)dataTaskForSystemRequestURLString:(NSString *)urlStrin
...
+ */
+- (NSURLSessionDownloadTask *)downloadTaskForSystemRequestURLString:(NSString *)urlString completionHandler:(URLSessionDownloadTaskCompletionHandler)completionHandler {
+    NSParameterAssert(urlString != nil);
+    NSParameterAssert(completionHandler != nil);
+    
+    NSString *logMessage = [NSString stringWithFormat:@"OnSystemRequest (HTTP Request to URL: %@", urlString];
+    [SDLDebugTool logInfo:logMessage withType:SDLDebugType_RPC toOutput:SDLDebugOutput_All toGroup:self.debugConsoleGroupName];
+    
+    // Create and return the download task
+    return [[NSURLSession sharedSession] downloadTaskWithURL:[NSURL URLWithString:urlString] completionHandler:completionHandler];
```
Comment (joeljfischer) [inline SmartDeviceLink-iOS/SmartDeviceLink/SDLProxy.m]: @adein This is iOS 7.0+, I'm assuming we should change it to a 6.0 compatible fetch?

Comment (adein) [inline SmartDeviceLink-iOS/SmartDeviceLink/SDLProxy.m]: @joeljfischer Yes, and Infosys will have a pull request soon that has the NSURLSession / NSURLConnection abstraction, which should work for this as well.

File: SmartDeviceLink-iOS/SmartDeviceLink/SDLProxy.m [hunk 3]
```diff
@@ -502,6 +505,32 @@ - (void)handleSystemRequestLaunchApp:(SDLOnSystemRequest *)request {
...
+        
+        if (error != nil) {
+            NSString *logMessage = [NSString stringWithFormat:@"OnSystemRequest failure (HTTP response), download task failed: %@", error.localizedDescription];
+            [SDLDebugTool logInfo:logMessage withType:SDLDebugType_RPC toOutput:SDLDebugOutput_All toGroup:self.debugConsoleGroupName];
+            return;
+        }
+        
+        UIImage *icon = [UIImage imageWithData: [NSData dataWithContentsOfURL:location]];
+        SEL selector = @selector(onReceivedLockScreenIcon:);
+        [self.proxyListeners enumerateObjectsUsingBlock:^(id listener, NSUInteger idx, BOOL *stop) {
```
Comment (joeljfischer) [inline SmartDeviceLink-iOS/SmartDeviceLink/SDLProxy.m]: see how this is made more clear in this PR: https://github.com/smartdevicelink/sdl_ios/blob/16fa7399dc1fe5430b878ed1bb06cee6aec9ce9d/SmartDeviceLink-iOS/SmartDeviceLink/SDLProxy.m#L627-634

File: SmartDeviceLink-iOS/SmartDeviceLink/SDLProxy.m (hunk 2 above)
Comment (joeljfischer) [inline SmartDeviceLink-iOS/SmartDeviceLink/SDLProxy.m]: A "download" feature may have to be added to the abstraction, I think it currently only handles data & upload.

**Comment by joeljfischer** [comment]
This is blocked by #290

**Comment by joeljfischer** [comment]
@adein This is done and tested on our side. Let me know when you've reviewed.

**Comment by joeljfischer** [comment]
@adein I lied, it isn't done yet.

**Comment by joeljfischer** [comment]
@adein now it's done

**Comment by adein** [comment]
LGTM :+1:

**Pull request merged**